### PR TITLE
Added skew detection functionality

### PR DESCRIPTION
**Pull request by kshivvy**
Added skew detection logic (using the serving_statistics argument in tfdv.validate_statistics) to the component executor.

### CODE REVIEW

**Comment by kshivvy** [comment]
@charlesccychen 
@yifanmai

File: tfx/types/standard_component_specs.py [hunk 1]
```diff
@@ -104,6 +104,8 @@ class ExampleValidatorSpec(ComponentSpec):
   INPUTS = {
       'statistics': ChannelParameter(type=standard_artifacts.ExampleStatistics),
       'schema': ChannelParameter(type=standard_artifacts.Schema),
+      'training_statistics':
+          ChannelParameter(type=standard_artifacts.ExampleStatistics),
```
Comment (charlesccychen) [inline tfx/types/standard_component_specs.py:108]: Can you make this an optional input?

File: tfx/components/example_validator/executor_test.py [hunk 2]
```diff
@@ -29,45 +29,72 @@
 
 class ExecutorTest(tf.test.TestCase):
 
-  def testDo(self):
-    source_data_dir = os.path.join(
-        os.path.dirname(os.path.dirname(__file__)), 'testdata')
+  def setUp(self):
+    super(ExecutorTest, self).setUp()
+    self.source_data_dir = os.path.join(
+        "../", os.path.dirname(os.path.dirname(__file__)), 'testdata')
```
Comment (charlesccychen) [inline tfx/components/example_validator/executor_test.py]: I don't think you want the prepended "../".

File: tfx/components/example_validator/executor_test.py [hunk 3]
```diff
@@ -29,45 +29,72 @@
 
 class ExecutorTest(tf.test.TestCase):
 
-  def testDo(self):
```
Comment (charlesccychen) [inline tfx/components/example_validator/executor_test.py:32]: Can you make sure this works with and without the optional argument?

**Comment by charlesccychen** [comment]
CC: @caveness @brills

File: tfx/components/example_validator/executor_test.py (hunk 2 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: You're right, I removed the "../"

File: tfx/types/standard_component_specs.py (hunk 1 above)
Comment (kshivvy) [reply tfx/types/standard_component_specs.py]: Made it an optional input

File: tfx/components/example_validator/executor_test.py (hunk 3 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py:32]: I added testDoSkewDetection() which uses the optional argument, and the existing testDo() does not use it, both versions seem to work. I also ended up extracting some common code into a setUp() and a helper method.

File: tfx/components/example_validator/executor.py [hunk 4]
```diff
@@ -36,6 +36,8 @@
 STATISTICS_KEY = 'statistics'
 # Key for schema in executor input_dict.
 SCHEMA_KEY = 'schema'
+# Key for training_statistics in executor input_dict.
```
Comment (yifanmai) [inline tfx/components/example_validator/executor.py:39]: nit: pre-existing whitespace is wonky; add a leading newline here and add one before line 37 too

File: tfx/components/example_validator/executor_test.py [hunk 5]
```diff
@@ -29,45 +29,72 @@
...
+    self.eval_stats_artifact.split_names = artifact_utils.encode_split_names(
         ['eval'])
 
-    schema_artifact = standard_artifacts.Schema()
-    schema_artifact.uri = os.path.join(source_data_dir, 'schema_gen')
+    self.schema_artifact = standard_artifacts.Schema()
+    self.schema_artifact.uri = os.path.join(self.source_data_dir, 'schema_gen')
 
-    output_data_dir = os.path.join(
+    self.output_data_dir = os.path.join(
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:46]: nit: `output_data_dir` is not used outside this method and can be local

File: tfx/components/example_validator/executor_test.py [hunk 6]
```diff
@@ -29,45 +29,72 @@
...
+        ['anomalies.pbtxt'],
+        tf.io.gfile.listdir(self.validation_output.uri))
     anomalies = io_utils.parse_pbtxt_file(
-        os.path.join(validation_output.uri, 'anomalies.pbtxt'),
+        os.path.join(self.validation_output.uri, 'anomalies.pbtxt'),
         anomalies_pb2.Anomalies())
     self.assertNotEqual(0, len(anomalies.anomaly_info))
+
+  def testDo(self):
+    self.input_dict = {
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:66]: `self.input_dict`, `self.output_dict` and `self.exec_properties` can all be locals, then you can pass them into `_do` as arguments. This makes the tests better encapsulated and more readable.

File: tfx/components/example_validator/executor_test.py [hunk 7]
```diff
@@ -29,45 +29,72 @@
...
     # TODO(zhitaoli): Add comparison to expected anomolies.
 
+  def testDoSkewDetection(self):
+    training_statistics = standard_artifacts.ExampleStatistics()
+    training_statistics.uri = os.path.join(
+        self.source_data_dir, 'trainer/current')
+    training_statistics.split_names = artifact_utils.encode_split_names(
+        ['eval_model_dir'])
+
+    self.input_dict = {
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:86]: These can be locals, as above.

File: tfx/components/example_validator/executor_test.py [hunk 8]
```diff
@@ -29,45 +29,72 @@
...
-    }
-    output_dict = {
-        executor.ANOMALIES_KEY: [validation_output],
-    }
-
-    exec_properties = {}
+    self.validation_output = standard_artifacts.ExampleAnomalies()
+    self.validation_output.uri = os.path.join(self.output_data_dir, 'output')
 
+  def _do(self):
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:53]: nit: rename to `_test_do` to make it obvious that the asserts actually happen within this method

File: tfx/components/example_validator/executor_test.py [hunk 9]
```diff
@@ -29,45 +29,72 @@
...
+    self.output_dict = {
+        executor.ANOMALIES_KEY: [self.validation_output],
+    }
+
+    self.exec_properties = {}
+
+    self._do()
     # TODO(zhitaoli): Add comparison to expected anomolies.
 
+  def testDoSkewDetection(self):
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:79]: The test as written does not test that skew detection produces an _additional_ anomaly due to skew: the test only asserts that the number of anomalies is non-zero, and the previous input already produces an anomaly. We could fix this by adding an assert that the `anomaly_info` contains an anomaly with the expected type(s).

File: tfx/components/example_validator/executor.py (hunk 4 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor.py:40]: Removed whitespace

File: tfx/components/example_validator/executor_test.py (hunk 5 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Done

File: tfx/components/example_validator/executor_test.py (hunk 6 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Sounds good, I made them all locals.

File: tfx/components/example_validator/executor_test.py (hunk 8 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Renamed to _test_do.

File: tfx/components/example_validator/executor_test.py (hunk 9 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py:89]: You were correct about this. I made sure that the anomalies detected for the skewed training set made sense by adding a skew_comparator for the 'payment_type' feature in the schema.pbtxt. I also added logic in the executor.py to make sure the training_statistics arguments were actually being parsed when passed in.

File: tfx/components/example_validator/executor_test.py (hunk 7 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Done.

File: tfx/components/example_validator/executor.py [hunk 10]
```diff
@@ -31,12 +31,15 @@
...
 
-
 # Key for statistics in executor input_dict.
 STATISTICS_KEY = 'statistics'
+
 # Key for schema in executor input_dict.
 SCHEMA_KEY = 'schema'
 
+# Key for training_statistics in executor input_dict.
+TRAINING_STATISTICS_KEY = 'training_statistics'
```
Comment (zhitaoli) [inline tfx/components/example_validator/executor.py:41]: nit: if these constants are only used in this module, consider privatize them.

For this and below.

File: tfx/components/example_validator/executor.py [hunk 11]
```diff
@@ -112,13 +121,18 @@ def _Validate(self, inputs: Dict[Text, Any], outputs: Dict[Text,
...
         - (Optional) labels.EXTERNAL_CONFIG_VERSION: the version number of
           external config file.
+        - (Optional) labels.TRAINING_STATISTICS: Training statistics to detect
+          skew against
       outputs: A dictionary of labeled output values, including:
           - labels.SCHEMA_DIFF_PATH: the path to write the schema diff to
     """
     schema = value_utils.GetSoleValue(inputs, labels.SCHEMA)
     stats = value_utils.GetSoleValue(inputs, labels.STATS)
+    serving_statistics = value_utils.GetSoleValue(
```
Comment (zhitaoli) [inline tfx/components/example_validator/executor.py:131]: Is this training or serving statistics? If the mismatch here is intentional, add a comment here to explain.

File: tfx/components/example_validator/executor_test.py [hunk 12]
```diff
@@ -19,55 +19,150 @@
...
+    p = pipeline.Pipeline(
+        pipeline_name='skewed_chicago_taxi_beam',
+        pipeline_root=tmp_dir,
+        components=[example_gen, statistics_gen, schema_gen],
+        metadata_connection_config=metadata.sqlite_metadata_connection_config(
+            os.path.join(tmp_dir, 'metadata.db')))
+
+    BeamDagRunner().run(p)
+
+    # Construct the skewed statsitics artifact from pipeline outputs
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:121]: typo: statistics

File: tfx/components/example_validator/executor_test.py [hunk 13]
```diff
@@ -19,55 +19,150 @@
...
+    original_schema_pbtxt_path = os.path.join(self.schema_artifact.uri,
+                                              'schema.pbtxt')
+    shutil.copy2(original_schema_pbtxt_path, tmp_schema_pbtxt_path)
+
+    f = open(tmp_schema_pbtxt_path, "r")
+    contents = f.readlines()
+    f.close()
+
+    skew_comparator_str = '  skew_comparator: { infinity_norm: { threshold: .1} }'
+    contents.insert(12, skew_comparator_str)
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:140]: This is brittle and could break if the line number changes... consider reading the file into proto, doing the modification in the proto, and then writing it back.

File: tfx/components/example_validator/executor_test.py [hunk 14]
```diff
@@ -19,55 +19,150 @@
...
+
+    # Set a skew comparator for the 'payment_type' feature
+    tmp_schema_pbtxt_dir = os.path.join(tmp_dir, 'SkewedSchemaGen')
+    os.mkdir(tmp_schema_pbtxt_dir)
+    tmp_schema_pbtxt_path = os.path.join(tmp_schema_pbtxt_dir, 'schema.pbtxt')
+    original_schema_pbtxt_path = os.path.join(self.schema_artifact.uri,
+                                              'schema.pbtxt')
+    shutil.copy2(original_schema_pbtxt_path, tmp_schema_pbtxt_path)
+
+    f = open(tmp_schema_pbtxt_path, "r")
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:135]: Prefer using `with open(tmp_schema_pbtxt_path, 'r') as f` instead; it's [more Pythonic](https://docs.python.org/3/tutorial/inputoutput.html#reading-and-writing-files). Same elsewhere.

File: tfx/components/example_validator/component.py [hunk 15]
```diff
@@ -84,6 +85,9 @@ def __init__(self,
         ExampleValidator. Required only if multiple ExampleValidator components
         are declared in the same pipeline.  Either `stats` or `statistics` must
         be present in the arguments.
+      training_statistics: Optional Channel of type
+        `standard_artifacts.ExampleStatistics` used for training-serving
+        skew detection
```
[the remark this answers is not in the document]
Comment (kshivvy) [reply tfx/components/example_validator/component.py]: Done.

File: tfx/components/example_validator/executor.py (hunk 11 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor.py]: The mismatch is intentional, I added a comment to explain.

File: tfx/components/example_validator/executor.py (hunk 10 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor.py:41]: I privatized `_DEFAULT_FILE_NAME`, but `TRAINING_STATISTICS_KEY` and `ANOMALIES_KEY` are used in `executor_test.py` so I didn't privatize them.

File: tfx/components/example_validator/executor_test.py (hunk 12 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Fixed typo

File: tfx/components/example_validator/executor_test.py (hunk 13 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Done. I read the file in proto, add the skew_comparator and then write it back.

File: tfx/components/example_validator/executor_test.py (hunk 14 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Done. I'll keep this in mind when I do file IO next time in python.

File: tfx/components/example_validator/executor_test.py [hunk 16]
```diff
@@ -19,55 +19,153 @@
...
+      proto_contents = text_format.Parse(contents, schema_pb2.Schema())
+
+      for feature in proto_contents.feature:
+        if feature.name == "payment_type":
+          feature.skew_comparator.infinity_norm.threshold = 0.001
+
+      contents = text_format.MessageToString(proto_contents)
+      f.seek(0)
+      f.write(contents)
+      f.truncate()
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:148]: IMHO it's cleaner to just reopen the file using "w", which truncates the file. (then you can open it using "r" the first time)

File: tfx/components/example_validator/executor_test.py [hunk 17]
```diff
@@ -19,55 +19,153 @@
...
+
+    output_dict = {
+        executor.ANOMALIES_KEY: [self.validation_output],
+    }
+
+    exec_properties = {}
+
+    # Generate anomalies and test that tfdv detected the skew
+    anomalies = self._test_do(input_dict, output_dict, exec_properties)
+    assert 'COMPARATOR_L_INFTY_HIGH' in str(anomalies.anomaly_info)
```
Comment (yifanmai) [inline tfx/components/example_validator/executor_test.py:166]: Use [`self.assertIn()`](https://docs.python.org/3/library/unittest.html#unittest.TestCase.assertIn). Note that the keyword assert is different from the assert methods.

File: tfx/components/example_validator/executor_test.py (hunk 16 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: I changed it to a second open with "w".

File: tfx/components/example_validator/executor_test.py (hunk 17 above)
Comment (kshivvy) [reply tfx/components/example_validator/executor_test.py]: Done, I'll use the prebuilt assert methods from now on.

**Comment by yifanmai** [comment]
LGTM. @charlesccychen could you take a look?

File: tfx/components/example_validator/component.py [hunk 18]
```diff
@@ -72,7 +72,8 @@ def __init__(self,
                exclude_splits: Optional[List[Text]] = None,
                output: Optional[types.Channel] = None,
                stats: Optional[types.Channel] = None,
-               instance_name: Optional[Text] = None):
+               instance_name: Optional[Text] = None,
+               training_statistics: Optional[types.Channel] = None):
```
Comment (charlesccychen) [inline tfx/components/example_validator/component.py:76]: Can you actually just rename this `serving_statistics` throughout?

**Comment by charlesccychen** [comment]
Please merge the PR https://github.com/kshivvy/tfx/pull/3 into this PR.

**Comment by martin-laurent** [comment]
Hey, why not using this PR to also add support for the "previous_statistics" and the "environment" parameter that are supported by TFDV but not TFX?

**Comment by martin-laurent** [comment]
This PR still looks very useful though. What is the next step?

**Pull request closed without merging**